### PR TITLE
Add support for custom static headers

### PR DESCRIPTION
**Pull request by vam-google**
Dependent on https://github.com/googleapis/gax-java/pull/432 (will fail build until that is released)

Note, PubSub is supported only partially. Full support is dependent on migration of PubSub from Grpc stubs to Gapic stubs. 

The spanner's `SpannerOptions.Builder.setUserAgentPrefix()` method is replaced with `ServiceOptions.Builder.setHeaderProvider()`  and providing "user-agent" value there (it will be prefixing the header provider as explained in `ServiceOptions.Builder.setHeaderProvider()` documentation, see example below).

Usage:
```java
// For manual/semi-manual clients:
//     Resource Manager
//     Translate
//     Compute
//     Storage
//     Datastore
//     Dns
//     BigQuery
//
//     Logging
//     Firestore
//     Spanner
HeaderProvider headerProvider =
    FixedHeaderProvider.create("user-agent", "my-client-name", "custom-header", "custom stuff");
DatastoreOptions options =
    DatastoreOptions.newBuilder().setHeaderProvider(headerProvider).build();
Datastore datastore = options.getService();

// For generated clients:
//    All others, except Pub/Sub
HeaderProvider headerProvider =
    FixedHeaderProvider.create("user-agent", "my-client-name", "custom-header", "custom stuff");
LanguageServiceSettings settings =
    LanguageServiceSettings.newBuilder().setHeaderProvider(headerProvider).build();
LanguageServiceClient client = LanguageServiceClient.create(settings);

// Note, PubSub is supported only partially. Full support is dependent on migration of PubSub from Grpc stubs to Gapic stubs
```

### CODE REVIEW

File: google-cloud-examples/src/main/java/com/google/cloud/examples/datastore/snippets/CreateEntity.java [hunk 1]
```diff
@@ -22,27 +22,38 @@
...
-/**
- * A snippet for Google Cloud Datastore showing how to create an entity.
- */
+/** A snippet for Google Cloud Datastore showing how to create an entity. */
 public class CreateEntity {
 
   public static void main(String... args) {
-    Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
+    HeaderProvider headerProvider =
+        FixedHeaderProvider.create("user-agent", "my-client-name", "custom-header", "custom stuff");
```
Comment (garrettjonesgoogle) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/datastore/snippets/CreateEntity.java]: I don't think it makes sense to include a custom header example in all of these samples...

Comment (vam-google) [reply google-cloud-examples/src/main/java/com/google/cloud/examples/datastore/snippets/CreateEntity.java]: Oh, that was accidental. This was just me testing this new functionality....

File: google-cloud-spanner/pom.xml [hunk 2]
```diff
@@ -18,38 +18,9 @@
...
-                   toFile="${project.build.outputDirectory}/com/google/cloud/spanner/project.properties"
-                   overwrite="true" />
-              </tasks>
-            </configuration>
-            <goals>
-              <goal>run</goal>
-            </goals>
-          </execution>
-        </executions>
-      </plugin>
```
Comment (vkedia) [inline google-cloud-spanner/pom.xml]: this seems unrelated?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java [hunk 3]
```diff
@@ -82,7 +84,7 @@ public ServiceRpc create(SpannerOptions options) {
   private SpannerOptions(Builder builder) {
     super(SpannerFactory.class, SpannerRpcFactory.class, builder, new SpannerDefaults());
     numChannels = builder.numChannels;
-    userAgent = builder.userAgentPrefix;
```
Comment (vkedia) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java]: what happens if the user also specifies this userAgentPrefix?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java [hunk 4]
```diff
@@ -97,25 +100,23 @@
...
-  private static final Metadata.Key<String> API_CLIENT_KEY =
-      Metadata.Key.of("x-goog-api-client", Metadata.ASCII_STRING_MARSHALLER);
-  private static final Metadata.Key<String> RESOURCE_PREFIX_KEY =
-      Metadata.Key.of("google-cloud-resource-prefix", Metadata.ASCII_STRING_MARSHALLER);
   private static final Pattern DATABASE_PATTERN =
-      Pattern.compile("^(?<database>projects/[^/]*/instances/[^/]*/databases/[^/]*)(.*)?");
+      Pattern.compile("^(?<headerValue>projects/[^/]*/instances/[^/]*/databases/[^/]*)(.*)?");
   private static final Pattern INSTANCE_PATTERN =
-      Pattern.compile("^(?<instance>projects/[^/]*/instances/[^/]*)(.*)?");
+      Pattern.compile("^(?<headerValue>projects/[^/]*/instances/[^/]*)(.*)?");
```
Comment (vkedia) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: why not just define these two in SpannerHeaderProvider and use there instead of passing from here?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerClientHeaderProvider.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+          Metadata.Key.of(entry.getKey(), Metadata.ASCII_STRING_MARSHALLER), entry.getValue());
+    }
+    return headersAsMetadataBuilder.build();
+  }
+
+  Map<Key<String>, String> getHeadersAsMetadata() {
+    return headersAsMetadata;
+  }
+
+  Map<Key<String>, String> getResourceHeadersAsMetadata(
```
Comment (vkedia) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerClientHeaderProvider.java]: this can be a static method.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerClientHeaderProvider.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+import com.google.api.gax.rpc.HeaderProvider;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class SpannerClientHeaderProvider implements HeaderProvider {
```
Comment (vkedia) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerClientHeaderProvider.java]: this implements `HeaderProvider` but I don't think `getHeaders` from this class is being called anywhere.
It might be better to just put constructHeadersAsMetadata and getResourceHeadersAsMetadata directly in GrpcSpannerRpc class

File: google-cloud-spanner/pom.xml (hunk 2 above)
Comment (vam-google) [reply google-cloud-spanner/pom.xml]: This is related. This change also simplifies the way `x-goog-api-client` header is handled differently now (just gets data from `METADATA.MF` and the project.properties is not needed anymore, thus removing its generation (same cleanup was done for all other poms in this repo).

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java (hunk 3 above)
Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java]: Good catch. That is a dead code now (user can specify userAgentPrefix in the builder, but it will be ignored). I've removed the rest of the related code.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java (hunk 4 above)
Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: This is possible. I was considering this too. Moved to SpannerHeaderProvider

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerClientHeaderProvider.java (hunk 6 above)
Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerClientHeaderProvider.java]: It is easier to test it if it is a separate class. Also I would like to keep headers stuff isolated, this has a good chance to be useful during Spanner GA work, when (if?) it gets migrated to gapic stubs.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerClientHeaderProvider.java (hunk 5 above)
Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerClientHeaderProvider.java]: It can be, but I'm trying to make spanner headers as similar in structure to general gapic headers logic for consistency and easier migration go gapic stubs in the future.

**Comment by vam-google** [comment]
@vkedia PTAL

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerClientHeaderProvider.java (hunk 6 above)
Comment (vkedia) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerClientHeaderProvider.java]: Ok, it can be a separate class but there is no need to implemented HeaderProvider for now. That can be added later if needed.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java (hunk 3 above)
Comment (vkedia) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java]: Can you give an example of how users would now specify customer user agent with spanner library? Ideally that should be prefixed to the default UA and not replace it completely.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java [hunk 7]
```diff
@@ -130,18 +125,19 @@ public GrpcSpannerRpc(SpannerOptions options) {
...
+    ApiClientHeaderProvider internalHeaderProvider =
+        internalHeaderProviderBuilder
+            .setClientLibToken(
+                ServiceOptions.getGoogApiClientLibName(),
+                GaxProperties.getLibraryVersion(options.getClass()))
+            .build();
+
+    HeaderProvider mergedHeaderProvider = options.getMergedHeaderProvider(internalHeaderProvider);
+    this.headerProvider = SpannerHeaderProvider.create(mergedHeaderProvider.getHeaders(),
+        internalHeaderProviderBuilder.getResourceHeaderKey());
```
Comment (vkedia) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: This key is a constant so I should be able to get to it by referring to a static constant rather than as a member method on ApiClientHeaderProvider. Ideally there should be no need to pass it to SpannerHeaderProvider. It should just refer to the constant somewhere.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerHeaderProvider.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+import com.google.api.gax.rpc.HeaderProvider;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class SpannerHeaderProvider implements HeaderProvider {
```
Comment (vkedia) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerHeaderProvider.java]: can  you add some class level doc.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerHeaderProvider.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    for (Pattern pattern : RESOURCE_TOKEN_PATTERNS) {
+      Matcher m = pattern.matcher(resourceTokenTemplate);
+      if (m.matches()) {
+        resourceToken = m.group("headerValue");
+        break;
+      }
+    }
+
+    return ImmutableMap.of(
+        Key.of(resourceHeaderKey, Metadata.ASCII_STRING_MARSHALLER), resourceToken);
```
Comment (vkedia) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerHeaderProvider.java]: This key can be a constant.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java [hunk 10]
```diff
@@ -474,19 +470,17 @@ public void start(Listener<RespT> responseListener, Metadata metadata) {
...
-    metadata.put(API_CLIENT_KEY, xGoogApiClientHeader);
-    return metadata;
-  }
 
-  @VisibleForTesting
-  String extractHeader(String resource) {
-    Matcher m = DATABASE_PATTERN.matcher(resource);
-    if (m.matches()) {
-      return m.group("database");
+    Map<Metadata.Key<String>, String> headers = headerProvider.getHeadersAsMetadata();
```
Comment (vkedia) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: I think it might be cleaner if this whole logic is put inside SpannerHeaderProvider. It can have a method like
void populateHeaders(Metadata metadata, String resource, String project);

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java (hunk 3 above)
Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java]: It is prefixed. In fact grpc itself does not allow to completely replace User-Agent: https://github.com/grpc/grpc-java/blob/c9b02db276403db4794c6e5ffc78b46889cd4ce8/netty/src/main/java/io/grpc/netty/Utils.java#L113

As for the example: it is exactly same as the examples in the original comment to this PR (the manual clients section). The provided User-Agent will be merged, not completelly replaced. It is also documented in the `ServiceOptions.setHeaderProvider()` method.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java (hunk 7 above)
Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: On gapic level this value is treated as the one with default value, but not constant. In this particular case, the value keeps the default value.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerHeaderProvider.java (hunk 8 above)
Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerHeaderProvider.java]: This is a package-private class.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerHeaderProvider.java (hunk 9 above)
Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerHeaderProvider.java]: Converted `SpannerHeaderProvider` to `SpannerMetadataProvider`, which now has only one method `newMetadata()`, with everything encapsulated there.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java (hunk 10 above)
Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GrpcSpannerRpc.java]: (same comment as above): Converted `SpannerHeaderProvider` to `SpannerMetadataProvider`, which now has only one method `newMetadata()`, with everything encapsulated there.

**Comment by vam-google** [comment]
@garrettjonesgoogle PTAL

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java (hunk 3 above)
Comment (vkedia) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java]: Is there some common documentation (not Javadoc) somewhere that we can point our customers to on how to set these headers.

Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java]: No. I hope it should not be more difficult than it was before this change. If somebody tries to set header he/she would be searching for some header setter method in the `SpannerOptions.Builder` (same as it was before) and will find `setHeadersProvider()` method (instead of `setUserAgentPrefix()`). It would be slightly less straightforward than before, but I believe should be still ok.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/SpannerMetadataProviderTest.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import java.util.Map;
+import org.junit.Test;
+
+public class SpannerMetadataProviderTest {
+  @Test
+  public void testGetHeadersAsMetadata() {
+    Map<String, String> headers = ImmutableMap.of("header1", "value1", "header2", "value2");
+    SpannerMetadataProvider headerProvider = SpannerMetadataProvider.create(headers, "header3");
```
Comment (garrettjonesgoogle) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/SpannerMetadataProviderTest.java]: rename the variable to match the type (`metadataProvider`)

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerMetadataProvider.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+package com.google.cloud.spanner.spi.v1;
+
+import com.google.common.collect.ImmutableMap;
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class SpannerMetadataProvider {
```
Comment (garrettjonesgoogle) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerMetadataProvider.java]: Add a comment about this being package-private

File: google-cloud-examples/src/main/java/com/google/cloud/examples/logging/snippets/CreateAndListMetrics.java [hunk 13]
```diff
@@ -35,7 +37,15 @@
...
     // Create a service object
     // Credentials are inferred from the environment
-    try(Logging logging = LoggingOptions.getDefaultInstance().getService()) {
+
+
+    HeaderProvider headerProvider =
+        FixedHeaderProvider.create("user-agent", "my-client-name", "custom-header", "custom stuff");
+
+    LoggingOptions options =
+        LoggingOptions.newBuilder().setHeaderProvider(headerProvider).build();
```
Comment (garrettjonesgoogle) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/logging/snippets/CreateAndListMetrics.java]: revert

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigquery/snippets/InsertDataAndQueryTable.java [hunk 14]
```diff
@@ -53,7 +53,7 @@ public static void main(String... args) throws InterruptedException {
     BigQuery bigquery = BigQueryOptions.getDefaultInstance().getService();
 
     // Create a dataset
-    String datasetId = "my_dataset_id";
+    String datasetId = "my_dataset_id_1";
```
Comment (garrettjonesgoogle) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/bigquery/snippets/InsertDataAndQueryTable.java]: revert

File: google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java [hunk 15]
```diff
@@ -535,40 +552,70 @@ public TransportOptions getTransportOptions() {
...
+   * Returns the application's name as a string in the format {@code gcloud-java/[version]},
+   * optionally prepended with externally supplied User-Agent header value (via setting custom
+   * header provider).
    */
   public String getApplicationName() {
     String libraryVersion = getLibraryVersion();
-    return libraryVersion == null ? LIBRARY_NAME : LIBRARY_NAME + "/" + libraryVersion;
-  }
 
+    // We have to do it, since underlying layers often do not appreciate User-Agent provided as a
```
Comment (garrettjonesgoogle) [inline google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java]: Better wording: "We have to do the following since ..."

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/SpannerMetadataProviderTest.java (hunk 11 above)
Comment (vam-google) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/SpannerMetadataProviderTest.java]: Done

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerMetadataProvider.java (hunk 12 above)
Comment (vam-google) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerMetadataProvider.java]: Done

File: google-cloud-examples/src/main/java/com/google/cloud/examples/logging/snippets/CreateAndListMetrics.java (hunk 13 above)
Comment (vam-google) [reply google-cloud-examples/src/main/java/com/google/cloud/examples/logging/snippets/CreateAndListMetrics.java]: Done

File: google-cloud-examples/src/main/java/com/google/cloud/examples/bigquery/snippets/InsertDataAndQueryTable.java (hunk 14 above)
Comment (vam-google) [reply google-cloud-examples/src/main/java/com/google/cloud/examples/bigquery/snippets/InsertDataAndQueryTable.java]: Done

File: google-cloud-bom/pom.xml [hunk 16]
```diff
@@ -14,8 +14,9 @@
     <api-client.version>1.23.0</api-client.version>
 
     <api-common.version>1.2.0</api-common.version>
-    <gax.version>1.15.0</gax.version>
-    <gax-grpc.version>1.15.0</gax-grpc.version>
+    <gax.version>1.15.1-SNAPSHOT</gax.version>
+    <gax-grpc.version>1.15.1-SNAPSHOT</gax-grpc.version>
+    <gax-httpjson.version>0.32.1-SNAPSHOT</gax-httpjson.version>
```
[the remark this answers is not in the document]
Comment (vam-google) [reply google-cloud-bom/pom.xml]: Will do

**Comment by vam-google** [comment]
@garrettjonesgoogle PTAL

**Comment by vam-google** [comment]
@garrettjonesgoogle PTAL
